Adding supported image formats in sm-list (XCP-ng 8.3)

gthvn1 opened the pull request:
This patch adds the field `supported-image-format` to SM object of the XAPI database. As it modifies the database we cannot have the same patch for our 8.3 and upstream because version of XAPI database or not the same. So upstream modifications conflicts with our modification. And the branch that follows the upstream one cannot be merged without conflicts.
So for testing and start review I created this patch that can be applied on our XCP-ng 8.3 branch so it is easy to test. Once reviewed I will rebase it upstream and see how we can do a PR.

gthvn1 committed:
Datamodel: add supported_image_format field to SM object

When running `xe sm-list params=all` you will now have the info of
supported image formats if the SM plugin specified it in its DRIVER_INFO.
The field is called `supported-image-formats`. If the plugin doesn't
provide the info the field will be empty.

This patch modifies the datamodel and add a new field to store this
information into the SM object.

Signed-off-by: Guillaume <guillaume.thouvenin@vates.tech>

last-genius reviewed (approved):
Are we confident about this? Or do we want to create a separate QCOW branch with this and other associated patches?

gthvn1 commented:
I think that it should be a default feature for XCP-ng as XO will display this information. And it will probably be used to allow the selection of a file format when the user creates a VDI. So I think that we don't need a specific branch for this. I mean we need to provide this feature.

gthvn1 commented on ocaml/xapi/xapi_vm_migrate.ml:
```diff
@@ -216,12 +217,76 @@ let assert_sr_support_operations ~__context ~vdi_map ~remote ~local_ops
+    supported image format of their corresponding SM. The type of the VDI is found in [vdi_format_map].
+    - If no VDI type is specified we just returned so no error is raised.
+    - If an SM reports an empty list of supported formats, we cannot verify compatibility and no error
+      is raised. So if the format is not actually supported, the failure will be detected later when
```
It maybe better to raise an error if the user pass a VDI type but SM doesn't have any types set. Currently it is accepted and so the creation of the VDI will fail. It is ok and the behavior is correct I think but we can also choose to just report an error during checks if no supported image formats are available and the user pass a type... Not sure what is the best approach.

last-genius replied:
The current behaviour sounds good to me

gthvn1 commented on ocaml/xapi/xapi_vm_migrate.ml:
```diff
@@ -216,12 +217,76 @@ let assert_sr_support_operations ~__context ~vdi_map ~remote ~local_ops
+                raise
+                  Api_errors.(
+                    Server_error
+                      ( vdi_incompatible_type
```
Currently we return incompatible type that makes sense I think but we can also create a new kind of error.

last-genius commented on ocaml/xapi-storage-cli/main.ml:
```diff
@@ -322,8 +322,11 @@ let mirror_start common_opts sr vdi dp url dest verify_dest =
       let dp = get_opt dp "Need a local data path" in
       let url = get_opt url "Need a URL" in
       let dest = get_opt dest "Need a destination SR" in
+      (* TODO: add image_format as a parameter of the CLI *)
```
Already done?

gthvn1 replied:
not yet...

last-genius commented on ocaml/xapi-cli-server/cli_operations.ml:
```diff
@@ -2060,7 +2060,13 @@ let vdi_pool_migrate printer rpc session_id params =
   and sr =
     Client.SR.get_by_uuid ~rpc ~session_id ~uuid:(List.assoc "sr-uuid" params)
-  and options = [] (* no options implemented yet *) in
+  and options =
+    [
+      List.assoc_opt "dest-img-format" params
+      |> Option.map (fun fmt -> ("dest_img_format", "&dest-img-format=" ^ fmt))
+    ]
+    |> List.filter_map (fun x -> x)
+  in
```
I'm not sure this is the best way of doing this. Maybe we could add `dest_img_format` to `VDI.pool_migrate` specified in `datamodel.ml`? This way we would avoid this annoying manual de/serialization, and get a proper versioned parameter

gthvn1 replied:
Yes agree. And I just updated the [PR for the design](https://github.com/xapi-project/xen-api/pull/6550) and I added a note saying that VDI.pool_migrate will have a new parameter. So it sounds good. I will do that.

last-genius reviewed (approved):
Aside from the comments, looks good to me overall. Would also be great if you could split the second commit into two:
1) `vdi_pool_migrate: Add the destination_format parameter`
2) `vm.migrate_send: Add the vdi_format_map  parameter`

These are independent pieces of work and it would be great to be able to revert them separately, if needed.

The first commit title could also be changed to something like:
```
datamodel: Add supported_image_formats field to the SM object
```
Since the fact that the new field shows up in `sm-list` is really a consequence of this

gthvn1 commented:
> Aside from the comments, looks good to me overall. Would also be great if you could split the second commit into two:
> 
>     1. `vdi_pool_migrate: Add the destination_format parameter`
> 
>     2. `vm.migrate_send: Add the vdi_format_map  parameter`

In fact I think I will split it in the other way because vdi.pool_migrate uses vm.migrate_send. So I think it makes more sense to start with vm migration and then vdi pool migrate.

gthvn1 commented:
Note: I set this PR back to draft because the goal, once we agree that it looks correct, is to rebase it upstream and propose it upstream. Here it is easier for testing and review to be based on XCP-ng 8.3 but the goal is to push it upstream at the end.

gthvn1 committed:
Allow selection of image format during VM migration

This patch allows specifying the destination format for individual VDIs
mapped to a destination SR. It adds a new parameter to `VM.migrate_send`
and `VM.assert_can_migrate` API. It also adds a new parameter to XE CLI.
The format to specify the image format is `image-format:<source VDI
UUID>=<destination image format>`. If the given image format cannot be
validated, an error is returned.

Signed-off-by: Guillaume <guillaume.thouvenin@vates.tech>

gthvn1 committed:
Allow selection of image format during VDI pool migration

This patch add a new parameter to `VDI.pool-migrate`. This new parameter
allows to provide a string that is the destination format. This string is used
to check whether the destination SR supports the expected format. If the check
fails or cannot be performed due to missing information on the destination SR,
an error is returned.

Signed-off-by: Guillaume <guillaume.thouvenin@vates.tech>

last-genius reviewed (approved):
Looks good to me. Could probably split the schema hash and datamodel version bump into a separate commit since they do not really only belong to the first one

last-genius commented:
And let's see what the upstream says :)

gthvn1 committed:
Bumping database schema version

A new field supported_image_format has been added to SM object and a new field
has been added to:
  - VM.migrate_send
  - VM.assert_can_migrate
  - VDI.pool_migrate

Signed-off-by: Guillaume <guillaume.thouvenin@vates.tech>

gthvn1 commented on ocaml/xapi/xapi_vm_migrate.ml:
```diff
@@ -216,12 +216,76 @@ let assert_sr_support_operations ~__context ~vdi_map ~remote ~local_ops
+              let sm_formats =
+                Db.SM.get_supported_image_formats ~__context ~self:sm_ref
+              in
+              if sm_formats <> [] && not (List.mem ty sm_formats) then
```
also need to check that **ty** is set because if no type is passed **ty** is empty string and so if supported image format is set (thus the list is not empty) the **List.mem** will fail.

gthvn1 committed:
fixup! Allow selection of image format during VM migration

gthvn1 commented:
When testing with qcow feature I notice an issue. I tried a migration from qcow2 to vhd and in fact only the snapshot is creating using VHD format. The mirror is still qcow2. So I need to also update the sm-config when the mirror is created.

gthvn1 committed:
fixup! Allow selection of image format during VM migration

gthvn1 committed:
fixup! Allow selection of image format during VM migration
